### PR TITLE
Various improvements for the JIT

### PR DESCRIPTION
**Pull request by apaszke**
* TensorType now records device information as well
* Fixed a minor bug in CppOp
* Fixed a bug where passing outputs to `save_for_backward` would break tracing (outputs wouldn't have the trace associated with them)
* Made certain Functions traceable. To complete the most important parts of `word_lanuage_model` we're still missing support for `Index`ing with slices (we should probably have our PyTorch custom index op) and `SubConstant` that I'm working on. I didn't manage to push it yet, because it needs to be implemented as Neg + AddConstant, but the code breaks with an assertion failure in the graph fuser if I don't set the type for Neg. I'll add support for manipulating types from Python and send a fix for this later.

### CODE REVIEW

File: torch/csrc/jit/ir.h [hunk 1]
```diff
@@ -87,37 +87,31 @@ struct Type {
...
-    strides_.resize(sizes_.size());
-    strides_.back() = 1;
-    for(size_t i = sizes_.size() - 1; i > 0; i--) {
-      strides_[i-1] = strides_[i]*sizes_[i];
-    }
-  }
+    : Type(TypeKind::TensorType)
+    , scalar_type_(tensor.type().scalarType())
+    , device_(tensor.type().isCuda() ? tensor.get_device() : -1)
+    , sizes_(tensor.sizes())
```
Comment (ezyang) [inline torch/csrc/jit/ir.h:93]: Heh. I always wondered why we needed that `std::copy`

File: torch/csrc/jit/ir.h [hunk 2]
```diff
@@ -87,37 +87,31 @@ struct Type {
...
+    t->strides_.resize(sizes_.size());
+    t->strides_.back() = 1;
+    for(size_t i = t->strides_.size() - 1; i > 0; i--) {
+      t->strides_[i-1] = t->strides_[i] * t->sizes_[i];
+    }
+    return t;
   }
 private:
   at::ScalarType scalar_type_;
+  int device_;
```
Comment (ezyang) [inline torch/csrc/jit/ir.h:114]: oh how I wish we had a struct for device id :) (you don't have to do anything about this comment)

**Comment by ezyang** [comment]
For the immediate PR, we need a test for SavedVariables tracing fix. We should also bump up the priority of the autograd/nn-like testing framework for tracing and running (but I wouldn't block the PR on this).

File: torch/autograd/function.py [hunk 3]
```diff
@@ -219,6 +222,11 @@ def err_fn(*args):
     return wrapper
 
 
+def traceable(fn_cls):
```
Comment (ezyang) [inline torch/autograd/function.py:225]: Needs a docstring

File: torch/csrc/autograd/python_function.cpp [hunk 4]
```diff
@@ -691,8 +701,9 @@ static void _trace_create(PyObject* op_obj, THPFunction* bw_obj,
   }
 
   // See definition in function.cpp.
-  // TODO: call into some Python method to check this
-  bool passes_state_transparently = false;
+  THPObjectPtr passes_py_bool {PyObject_GetAttrString(op_obj, "is_traceable")};
+  if (!passes_py_bool) throw python_error();
+  bool passes_state_transparently = passes_py_bool == Py_True;
```
Comment (ezyang) [inline torch/csrc/autograd/python_function.cpp:703]: Shouldn't we use `PyObject_IsTrue` here?

**Comment by ezyang** [comment]
Thanks. Tests need to pass before going in.

File: torch/autograd/function.py (hunk 3 above)
Comment (apaszke) [reply torch/autograd/function.py:225]: It's only for internal purposes for now, and I'm not sure if this API is going to be public any time.

**Pull request merged**